### PR TITLE
Re-enable the ai review action with a few conditions

### PR DESCRIPTION
**Pull request by jimleroyer**
# Summary | Résumé

This change re-enables the AI code review github action but changes its trigger. Whereas before, this triggered on new pull request or updates to it, this would now trigger on a commit that contains the `[review]` keyword in the description.

> [!CAUTION]
> This might break the script as it expects a github event of `opened` or `synchronize`. If that breaks it, we need to look how much changes that would require and implement these.

> **Note**: Ajoutez vous une nouvelle page par Articles GC? Assurez vous d'ajouter le chemin dans les [règles WAF](https://github.com/cds-snc/notification-utils/tree/main/.github/actions/waffles#supporting-a-new-url-within-gcnotify).

# Test instructions | Instructions pour tester la modification

The job was skipped on this PR as no commit contained the keyword `[review]`:
image

### CODE REVIEW

File: .github/workflows/code_review.yml [hunk 1]
```diff
@@ -1,12 +1,14 @@
 name: CDS Code Review with OpenAI
-# on:
-#   pull_request:
-#     types:
-#       - opened
-#       - synchronize
+on:
+  push:
```
[the remark this answers is not in the document]
Comment (jimleroyer) [reply .github/workflows/code_review.yml:3]: We could keep it on PR but we could not filter on the `[request]` keyword found in commits then. Only commits with this keyword would trigger the review, hence the filter that would reduce the number of runs. 

As an alternative, we could just trigger this on `opened` event (new PR) and guarantee one run but prevent other ones.

**Review by ben851 — APPROVED** [review]
LGTM

File: src/main.ts [hunk 2]
```diff
@@ -67,10 +71,49 @@ interface PRDetails {
...
+ *
+ * @param {GitHubEvent} event - The type of GitHub event ("opened", "synchronize", or "push").
+ * @returns {Promise<PRDetails>} - A promise that resolves to the pull request details.
+ *
+ * @throws {Error} - Throws an error if the event type is unsupported.
+ */
+async function getPrDetails(
+  eventName: GitHubEvent,
+  eventData: any
+): Promise<PRDetails> {
```
[the remark this answers is not in the document]
Comment (jimleroyer) [reply src/main.ts:89]: We could, but we do not have the type wrappers for Github events AFAIK.

File: src/main.ts [hunk 3]
```diff
@@ -85,6 +128,50 @@ async function getPRDetails(): Promise<PRDetails> {
...
 }
 
+/**
+ * Retrieves the pull request details associated with a given push event from the GitHub API.
+ *
+ * @param eventData - The event data containing information about the push event.
+ * @returns A promise that resolves to the details of the associated pull request.
+ * @throws Will throw an error if no associated pull request is found for the given push event.
+ */
+async function getPrFromApi(eventData: any): Promise<PRDetails> {
```
[the remark this answers is not in the document]
Comment (jimleroyer) [reply src/main.ts:138]: We could, but we do not have the type wrappers for Github events AFAIK.

File: src/main.ts [hunk 4]
```diff
@@ -235,63 +360,51 @@ function createComment(
...
+ *
+ * This function reads the `exclude` and `include` patterns from the GitHub Action inputs,
+ * trims and filters out any empty strings, and then applies these patterns to the parsed diff files.
+ * Files that match the exclude patterns are excluded, and files that match the include patterns are included.
+ * If both patterns are provided, the exclude patterns take precedence over the include patterns.
+ *
+ * @param {File[]} parsedDiff - An array of parsed diff files to be filtered.
+ * @returns {File[]} - An array of filtered diff files.
+ */
+function filterDiffs(parsedDiff: File[]): File[] {
```
[the remark this answers is not in the document]
Comment (jimleroyer) [reply src/main.ts:407]: It is imported.

File: src/main.ts [hunk 5]
```diff
@@ -317,8 +430,69 @@ async function main() {
     return !excluded && included;
   });
 
+  return filteredDiff;
+}
+
+async function main() {
```
[the remark this answers is not in the document]
Comment (jimleroyer) [reply src/main.ts:436]: It is not defined twice.

File: src/main.ts [hunk 6]
```diff
@@ -317,8 +430,69 @@ async function main() {
...
+  const eventData = JSON.parse(
+    readFileSync(process.env.GITHUB_EVENT_PATH ?? "", "utf8")
```
[the remark this answers is not in the document]
Comment (jimleroyer) [reply src/main.ts:438]: Not a bad review but there is a global error handler defined at the `main` level.

File: src/main.ts [hunk 7]
```diff
@@ -317,8 +430,69 @@ async function main() {
...
+        prDetails.owner,
+        prDetails.repo,
+        prDetails.pull_number
+      );
+      break;
+    case "synchronize":
+      const newBaseSha = eventData.before;
+      const newHeadSha = eventData.after;
+
+      const response = await octokit.repos.compareCommits({
```
[the remark this answers is not in the document]
Comment (jimleroyer) [reply src/main.ts:462]: It is defined higher in the script, but the PR does not contain its declaration. It is used globally, which is a smell. Maybe a better comment would be to mention to not use such global instances.

File: src/main.ts [hunk 8]
```diff
@@ -317,8 +430,69 @@ async function main() {
...
+    console.log("No diff found");
+    return;
+  }
+
+  const parsedDiff = parseDiff(diff);
+  const filteredDiff = filterDiffs(parsedDiff);
+
   const comments = await analyzeCode(filteredDiff, prDetails);
   if (comments.length > 0) {
+    // Additional logging and validation before creating the review
```
[the remark this answers is not in the document]
Comment (jimleroyer) [reply src/main.ts:489]: Not the best comment to leave in code but explaining code should be a good thing.

**Review by sastels — APPROVED** [review]
LGTM 🚀

**Pull request merged**